### PR TITLE
Prevent panic when sysroot cannot be computed

### LINKED ISSUE #61377: rustdoc should not assume `current_exe()` is infallible
On OpenBSD, the `std::env::current_exe()` function fails if the binary is not invoked with a full path (see [this comment](https://github.com/rust-lang/rust/pull/21964#issue-28730541)). This leads to the following crash when running doc tests:

```
% cargo test
[...]
   Doc-tests zmq
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: Custom { kind: Other, error: StringError("no current exe available (short)") }', src/libcore/result.rs:997:5
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Note that this using rustc/cargo 1.33.0, installed on OpenBSD 6.5 using its native package (`pkg_add rust`).

The offending code, I believe, must be one of the `current_exe()` calls in `src/librustdoc/test.rs`, which are both unwrapped, and have to do with figuring out the sysroot, if not already known.

The issue can be worked around by adjusting the command line like this:

```
% RUSTDOCFLAGS='--sysroot /usr/local' cargo test
```
I guess solving this completely may be tricky, but I think the code should definitely not panic, and ideally cargo/rustdoc would issue an error message pointing (OpenBSD) users to the workaround.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #61377.

cc @rotty @rust-lang/rustdoc 

r? @Manishearth

### CODE REVIEW

File: src/librustdoc/test.rs [hunk 1]
```diff
@@ -40,12 +40,26 @@ pub struct TestOptions {
...
     let sessopts = config::Options {
-        maybe_sysroot: options.maybe_sysroot.clone().or_else(
-            || Some(env::current_exe().unwrap().parent().unwrap().parent().unwrap().to_path_buf())),
+        maybe_sysroot: options.maybe_sysroot.clone().or_else(|| {
+            Some(try_none!(
+                try_none!(
+                    try_none!(env::current_exe().ok()).parent()
+                ).parent()
+            ).to_path_buf())
+        }),
```
Comment (ollie27) [inline src/librustdoc/test.rs]: This should just be `maybe_sysroot: options.maybe_sysroot.clone(),` which will let rustc find the sysroot if it wasn't passed as a command line attribute. Unless there's some reason rustdoc needs to call `current_exe` itself but I can't see any.

File: src/librustdoc/test.rs [hunk 2]
```diff
@@ -225,8 +239,13 @@ fn run_test(
...
     let sessopts = config::Options {
-        maybe_sysroot: maybe_sysroot.or_else(
-            || Some(env::current_exe().unwrap().parent().unwrap().parent().unwrap().to_path_buf())),
+        maybe_sysroot: maybe_sysroot.or_else(|| {
+            Some(try_none!(
+                try_none!(
+                    try_none!(env::current_exe().ok()).parent()
+                ).parent()
+            ).to_path_buf())
+        }),
```
Comment (ollie27) [inline src/librustdoc/test.rs]: Again, this can just be `maybe_sysroot,`.

**Review by ollie27 — CHANGES_REQUESTED** [review]

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/librustdoc/test.rs [hunk 3]
```diff
@@ -225,8 +224,7 @@ fn run_test(
     let outputs = OutputTypes::new(&[(OutputType::Exe, None)]);
 
     let sessopts = config::Options {
-        maybe_sysroot: maybe_sysroot.or_else(
-            || Some(env::current_exe().unwrap().parent().unwrap().parent().unwrap().to_path_buf())),
+        maybe_sysroot: maybe_sysroot,
```
Comment (bjorn3) [inline src/librustdoc/test.rs]:
```suggestion
        maybe_sysroot,
```

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by QuietMisdreavus — APPROVED** [review]
LGTM! Looks like passing `None` all the way down to rustc lets it use the proper calculations on OpenBSD.

I did some archaeology to see the history of the sysroot parameter. It looks like the `current_exe` logic was taken out of `core.rs` way back in https://github.com/rust-lang/rust/pull/19161, but that PR didn't touch `test.rs`, and the logic was carried through during the various IO refactorings over time. This brings the doctest functionality in-line with the rest of rustdoc, and the rest of rustc.

We don't have OpenBSD builders, do we? I'm wondering if there's a way we can test for this. 😕

**Comment by GuillaumeGomez** [comment]
@rust-lang/infra Can we check on OpenBSD one way or another?

**Comment by GuillaumeGomez** [comment]
Well, meanwhile, this PR is ready to go.

@bors: r=ollie27,bjorn3,QuietMisdreavus

**Comment by GuillaumeGomez** [comment]
Maybe it'll work this time?

@bors: r=ollie27,bjorn3,QuietMisdreavus

**Comment by bjorn3** [comment]
@GuillaumeGomez Why did you include me in the `r=` list? Every other time I commented on a PR, I didnt get included.

**Comment by GuillaumeGomez** [comment]
Well, you reviewed it so it seemed ok for me to include you.

**Pull request merged**